

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>libPomo.cf &mdash; PoMo 1.0.2 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="PoMo 1.0.2 documentation" href="index.html" />
    <link rel="up" title="libPoMo" href="libPoMo.html" />
    <link rel="prev" title="libPomo.vcf" href="vcf.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="vcf.html" title="libPomo.vcf"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">PoMo 1.0.2 documentation</a> &raquo;</li>
          <li><a href="libPoMo.html" accesskey="U">libPoMo</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="module-libPoMo.cf"></span><div class="section" id="libpomo-cf">
<h1>libPomo.cf<a class="headerlink" href="#libpomo-cf" title="Permalink to this headline">¶</a></h1>
<p>This model provides functions to read, write and access files that are
in counts format.</p>
<div class="section" id="the-counts-format">
<h2>The Counts Format<a class="headerlink" href="#the-counts-format" title="Permalink to this headline">¶</a></h2>
<p>This file format is used by PoMo and lists the base
counts for every position.</p>
<dl class="docutils">
<dt>It contains:</dt>
<dd><ul class="first last simple">
<li>1 headerline with tab separated sequence names</li>
<li>N lines with counts of A, C, G and T bases at position n</li>
</ul>
</dd>
</dl>
<div class="highlight-python"><pre>CHROM          Pos     Sheep           BlackSheep      RedSheep        Wolf            RedWolf
chrA           s       0,0,1,0         0,0,1,0         0,0,1,0         0,0,5,0         0,0,0,1
chrA           s + 1   0,0,0,1         0,0,0,1         0,0,0,1         0,0,0,5         0,0,0,1
.
.
.
chrF           8373    0,0,0,1         1,0,0,0         0,1,0,0         0,1,4,0         0,0,1,0
.
.
.
chrE           end     0,0,0,1         0,1,0,0         0,1,0,0         0,5,0,0         0,0,1,0</pre>
</div>
</div>
<div class="section" id="objects">
<h2>Objects<a class="headerlink" href="#objects" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>Classes:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#libPoMo.cf.CFWriter" title="libPoMo.cf.CFWriter"><tt class="xref py py-class docutils literal"><span class="pre">CFWriter</span></tt></a>, write a counts format file</li>
</ul>
</dd>
<dt>Exception Classes:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#libPoMo.cf.CountsFormatWriterError" title="libPoMo.cf.CountsFormatWriterError"><tt class="xref py py-class docutils literal"><span class="pre">CountsFormatWriterError</span></tt></a></li>
</ul>
</dd>
<dt>Functions:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#libPoMo.cf.save_as_cf" title="libPoMo.cf.save_as_cf"><tt class="xref py py-func docutils literal"><span class="pre">save_as_cf()</span></tt></a>, deprecated; save given sequences to a counts
format file</li>
</ul>
</dd>
</dl>
<hr class="docutils" />
<dl class="class">
<dt id="libPoMo.cf.CFWriter">
<em class="property">class </em><tt class="descclassname">libPoMo.cf.</tt><tt class="descname">CFWriter</tt><big>(</big><em>vcfFileNameL</em>, <em>outFileName</em>, <em>verb=None</em>, <em>mergeL=None</em>, <em>nameL=None</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#CFWriter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.CFWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a counts format file.</p>
<p>Save information that is needed to write a CF file and use this
information to write a CF file.  Initialize with a list of vcf
file names and an output file name:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">CFWriter</span><span class="p">([</span><span class="n">vcfFileNames</span><span class="p">],</span> <span class="s">&quot;output&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Before the count file can be written, a reference sequence has to
be specified.  A single reference sequence can be set with
<a class="reference internal" href="#libPoMo.cf.CFWriter.set_seq" title="libPoMo.cf.CFWriter.set_seq"><tt class="xref py py-class docutils literal"><span class="pre">set_seq</span></tt></a>.</p>
<p>Write a header line to output:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="bp">self</span><span class="o">.</span><span class="n">write_HLn</span><span class="p">()</span>
</pre></div>
</div>
<p>Write lines in counts format from 1-based positions <em>start</em> to
<em>end</em> on chromosome <em>chrom</em> to output:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">rg</span> <span class="o">=</span> <span class="n">sb</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="s">&quot;chrom&quot;</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
<span class="bp">self</span><span class="o">.</span><span class="n">write_Rn</span><span class="p">(</span><span class="n">rg</span><span class="p">)</span>
</pre></div>
</div>
<p>If you want to compare the SNPs of the VCF files to a multiple
alingment fasta stream (<a class="reference internal" href="fasta.html#libPoMo.fasta.MFaStream" title="libPoMo.fasta.MFaStream"><tt class="xref py py-class docutils literal"><span class="pre">MFaStream</span></tt></a>) consider the very convenient function
<a class="reference internal" href="#libPoMo.cf.write_cf_from_MFaStream" title="libPoMo.cf.write_cf_from_MFaStream"><tt class="xref py py-func docutils literal"><span class="pre">write_cf_from_MFaStream()</span></tt></a>.</p>
<p>Remember to close the attached file objectsL with <a class="reference internal" href="#libPoMo.cf.CFWriter.close" title="libPoMo.cf.CFWriter.close"><tt class="xref py py-func docutils literal"><span class="pre">close()</span></tt></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>vcfFileNameL</strong> (<em>[str]</em>) &#8211; List with names of vcf files.</li>
<li><strong>outFileName</strong> (<em>str</em>) &#8211; Output file name.</li>
<li><strong>verb</strong> (<em>int</em>) &#8211; Optional; verbosity.</li>
<li><strong>mergeL</strong> (<em>[Boolean]</em>) &#8211; Optional; a list of truth values.  If
<em>mL[i]</em> is True, all individuals of <em>self.vcfL[i]</em> are treated as
one population orspecies independent of their name.  The
respective counts are summed up.  If <em>self.nL[i]</em> is given, the
name of the summed sequence will be <em>self.nL[i]</em>.  If not, the
name of the first individual in <em>vcfL[i]</em> will be used.</li>
<li><strong>nameL</strong> (<em>[str]</em>) &#8211; Optional; a list of names. Cf. <em>self.mL</em>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>refFN</strong> (<em>str</em>) &#8211; Name of reference fasta file.</li>
<li><strong>vcfL</strong> (<em>[str]</em>) &#8211; List with names of vcf files.</li>
<li><strong>outFN</strong> (<em>str</em>) &#8211; Output file name.</li>
<li><strong>v</strong> (<em>int</em>) &#8211; Verbosity.</li>
<li><strong>mL</strong> (<em>[Boolean]</em>) &#8211; A list of truth values.  If <em>mL[i]</em> is True,
all individuals of <em>self.vcfL[i]</em> are treated as one
population orspecies independent of their name.  The
respective counts are summed up.  If <em>self.nL[i]</em> is given,
the name of the summed sequence will be <em>self.nL[i]</em>.  If not,
the name of the first individual in <em>vcfL[i]</em> will be used.</li>
<li><strong>nL</strong> (<em>[str]</em>) &#8211; A list of names. Cf. <em>self.mL</em>.</li>
<li><strong>nV</strong> (<em>int</em>) &#8211; Number of vcf files.</li>
<li><strong>vcfTfL</strong> (<em>[fo]</em>) &#8211; List with <em>pysam.Tabixfile</em> objects. Filled by
<em>self.__init_vcfTfL()</em> during initialization.</li>
<li><strong>outFO</strong> (<em>fo</em>) &#8211; File object of the outfile. Filled by
<em>self.__init_outFO()</em> during initialization.</li>
<li><strong>cD</strong> &#8211; List with allele or base counts. The alleles of
individuals from the same population are summed up.  Hence,
<em>self.cD[p]</em> gives the base counts of population <em>p</em> in the
form: [0, 0, 0, 0].  Population <em>p`*does not need to be the
one from *self.vcfL[p]</em> because several populations might be
present in one vcf file.  <em>self.assM</em> connects the individual
j from <em>self.vcfL[i]</em> such that <em>self.assM[i][j]</em> is <em>p</em>.</li>
<li><strong>chrom</strong> (<em>str</em>) &#8211; Name of the current chromosome. Set and updated
by <a class="reference internal" href="#libPoMo.cf.CFWriter.write_Rn" title="libPoMo.cf.CFWriter.write_Rn"><tt class="xref py py-func docutils literal"><span class="pre">write_Rn()</span></tt></a>.</li>
<li><strong>pos</strong> (<em>int</em>) &#8211; Current position on chromosome. Set and updated by
<a class="reference internal" href="#libPoMo.cf.CFWriter.write_Rn" title="libPoMo.cf.CFWriter.write_Rn"><tt class="xref py py-func docutils literal"><span class="pre">write_Rn()</span></tt></a>.</li>
<li><strong>offset</strong> (<em>int</em>) &#8211; Value that can be set with <a class="reference internal" href="#libPoMo.cf.CFWriter.set_offset" title="libPoMo.cf.CFWriter.set_offset"><tt class="xref py py-func docutils literal"><span class="pre">set_offset()</span></tt></a>,
if the reference sequence does not start at the
1-based position 1 but at the 1-based position
<em>offset</em>.</li>
<li><strong>indM</strong> &#8211; Matrix with individuals from vcf files. <em>self.indM[i]</em>
is the list of individuals found in <em>self.vcfL[i]</em>.</li>
<li><strong>nIndL</strong> (<em>[int]</em>) &#8211; List with number of individuals in
<em>self.vcfL[i]</em>.</li>
<li><strong>assM</strong> &#8211; Assignment matrix that connects the individuals from
the vcf files to the correct <em>self.cd</em> index.  Cf. <em>self.cD</em></li>
<li><strong>nPop</strong> (<em>int</em>) &#8211; Number of different populations in count format
output file (e.g. number of populations).  Filled by
<em>self.__init_assM()</em> during initialization.</li>
<li><strong>refSeq</strong> (<em>Seq</em>) &#8211; <a class="reference internal" href="seqbase.html#libPoMo.seqbase.Seq" title="libPoMo.seqbase.Seq"><tt class="xref py py-class docutils literal"><span class="pre">Seq</span></tt></a> object of the
reference Sequence. This has to be set with <a class="reference internal" href="#libPoMo.cf.CFWriter.set_seq" title="libPoMo.cf.CFWriter.set_seq"><tt class="xref py py-class docutils literal"><span class="pre">set_seq</span></tt></a>.</li>
<li><strong>ploidy</strong> (<em>int</em>) &#8211; Ploidy of individuals in vcf files.  This has to
be set manually to the correct value for non-diploids!</li>
<li><strong>__splitCh</strong> (<em>char</em>) &#8211; Character that is used to split the
individual names.</li>
<li><strong>__force</strong> (<em>Boolean</em>) &#8211; If set to true, skip name checks.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="function">
<dt id="libPoMo.cf.CFWriter.close">
<tt class="descname">close</tt><big>(</big><em>self</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#CFWriter.close"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.CFWriter.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close fileobjects.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.CFWriter.set_force">
<tt class="descname">set_force</tt><big>(</big><em>self</em>, <em>val</em><big>)</big><a class="headerlink" href="#libPoMo.cf.CFWriter.set_force" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets <em>self.__force</em> to <em>val</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>val</strong> (<em>Boolean</em>) &#8211; </td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.CFWriter.set_offset">
<tt class="descname">set_offset</tt><big>(</big><em>self</em>, <em>offset</em><big>)</big><a class="headerlink" href="#libPoMo.cf.CFWriter.set_offset" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the offset of the sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>offset</strong> (<em>int</em>) &#8211; Value that can be set, if the reference
sequence does not start at the 1-based
position 1 but at the 1-based position
<em>offset</em>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.CFWriter.set_seq">
<tt class="descname">set_seq</tt><big>(</big><em>self</em>, <em>seq</em><big>)</big><a class="headerlink" href="#libPoMo.cf.CFWriter.set_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the reference sequence.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.CFWriter.write_HLn">
<tt class="descname">write_HLn</tt><big>(</big><em>self</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#CFWriter.write_HLn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.CFWriter.write_HLn" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the counts format header line to <em>self.outFN</em>.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.CFWriter.write_Rn">
<tt class="descname">write_Rn</tt><big>(</big><em>self</em>, <em>rg</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#CFWriter.write_Rn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.CFWriter.write_Rn" title="Permalink to this definition">¶</a></dt>
<dd><p>Write lines in counts format to <em>self.outFN</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>rg</strong> (<em>Region</em>) &#8211; <a class="reference internal" href="seqbase.html#libPoMo.seqbase.Region" title="libPoMo.seqbase.Region"><tt class="xref py py-class docutils literal"><span class="pre">Region</span></tt></a>
object that determines the region that is
covered.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="libPoMo.cf.CountsFormatWriterError">
<em class="property">exception </em><tt class="descclassname">libPoMo.cf.</tt><tt class="descname">CountsFormatWriterError</tt><a class="reference internal" href="_modules/libPoMo/cf.html#CountsFormatWriterError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.CountsFormatWriterError" title="Permalink to this definition">¶</a></dt>
<dd><p>General <cite>CFWriter</cite> object error.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.collapse">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">collapse</tt><big>(</big><em>speciesL</em>, <em>merge=False</em>, <em>name=None</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#collapse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.collapse" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Collapse the species names.</p>
<p>Collapse the species names using the naming rules given in
save_as_countsformat. Returns a dictionary with collapsed
species names as keys and an assignment list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>merge</strong> (<em>Boolean</em>) &#8211; if set to True, all species/individuals will be
collapsed to a single one with name <em>name</em> (if <em>name</em> is given).</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.fill_species_dict_alt">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">fill_species_dict_alt</tt><big>(</big><em>spDi</em>, <em>assL</em>, <em>refBase</em>, <em>altBases</em>, <em>spData</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#fill_species_dict_alt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.fill_species_dict_alt" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Fill the species dictionary when a SNP is present.</p>
<p>Raise <em>SequenceDataError</em>, if <em>spDi</em> could not be filled.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.fill_species_dict_ref">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">fill_species_dict_ref</tt><big>(</big><em>spDi</em>, <em>assL</em>, <em>refBase</em>, <em>ploidy</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#fill_species_dict_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.fill_species_dict_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Fills the species dictionary if no SNP is present.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.find_next_SNP_pos">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">find_next_SNP_pos</tt><big>(</big><em>nSNPChromL</em>, <em>nSNPPosL</em>, <em>ref</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#find_next_SNP_pos"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.find_next_SNP_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Find the position of next SNP.</p>
<p>Return the position of the next SNP in <em>ref</em> and a list of
indices of these SNPs in <em>nSNPChromL</em> and <em>nSNPPosL</em>. If no
next SNP is found on this sequence (this might happen if all
next SNPs are on the next chromosome) a ValueError is raised.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>nSNPChromL</strong> &#8211; List with chromosome names of the next SNPs of
the <em>vcfStrL</em>.</li>
<li><strong>nSNPPosL</strong> &#8211; List with positions of the next SNPs of the
<em>vcfStrL</em>.</li>
<li><strong>ref</strong> &#8211; Seq object of the sequence of the reference.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.get_cf_headerline">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">get_cf_headerline</tt><big>(</big><em>species</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#get_cf_headerline"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.get_cf_headerline" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Return a string containing the headerline in counts
format.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.get_counts_line">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">get_counts_line</tt><big>(</big><em>chrom</em>, <em>pos</em>, <em>spDi</em>, <em>speciesL</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#get_counts_line"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.get_counts_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Return line string in counts format.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.save_as_cf">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">save_as_cf</tt><big>(</big><em>vcfStrL</em>, <em>refFaStr</em>, <em>CFFileName</em>, <em>verb=False</em>, <em>mergeL=None</em>, <em>nameL=None</em><big>)</big><a class="reference internal" href="_modules/libPoMo/cf.html#save_as_cf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.cf.save_as_cf" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated. Save the given sequence in counts format.</p>
<p>This function saves the SNPs from <em>vcfStrL</em>, a given list of
<em>VCFStream</em> (variant call format sequence stream) objects in
counts format to the file <em>CFFileName</em>.  The reference genome
<em>refFaStr</em>, to which <em>VCFSeqStr</em> is compared to, needs to be
passed as an <em>FaStream</em> object.</p>
<p>The name of all streams in <em>vcfStrL</em> should be the same as the
name of <em>faRef</em>.  The names of the sequences in the reference
should be the names of the chromosomes found in the <em>vcfStr</em>
object, otherwise we do not know where to compare the sequences
to.  They must also be in the same order!</p>
<p>Individuals with the same name and suffix &#8220;_n&#8221;, where n is a
number, will be saved in one column without the suffix.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>vcfStrL</strong> &#8211; List with VCF Streams containing SNPs.</li>
<li><strong>refFaStr</strong> &#8211; Reference fasta sequence stream object.</li>
<li><strong>CFFileName</strong> &#8211; Name of output file (counts format).</li>
<li><strong>verb</strong> &#8211; If <em>verb</em> is set to True, additional information is
printed to the output file.</li>
<li><strong>mergeL</strong> &#8211; A list of truth values. If <em>mergeL[i]</em> is True, all
individuals of <em>vcfStrL[i]</em> are treated as one species
independent of their name. The respective counts are
summed up.  If <em>nameL[i]</em> is given, the name of the
summed sequence will be <em>nameL[i]</em>. If not, the name of
the first individual in <em>vcfStrL[i]</em> will be used.</li>
<li><strong>nameL</strong> &#8211; A list of names. Cf. <em>mergeL</em>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.cf.write_cf_from_MFaStream">
<tt class="descclassname">libPoMo.cf.</tt><tt class="descname">write_cf_from_MFaStream</tt><big>(</big><em>refMFaStr</em>, <em>cfWr</em><big>)</big><a class="headerlink" href="#libPoMo.cf.write_cf_from_MFaStream" title="Permalink to this definition">¶</a></dt>
<dd><p>Write counts file using the given MFaStream and CFWriter.</p>
<p>Write the counts format file using the first sequences of all
alignments in MFaStream.  This is very useful if you e.g. want
to compare the VCF files to a CCDC alignment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>refMFaStr</strong> (<em>FMaStream</em>) &#8211; The reference <a class="reference internal" href="fasta.html#libPoMo.fasta.MFaStream" title="libPoMo.fasta.MFaStream"><tt class="xref py py-class docutils literal"><span class="pre">MFaStream</span></tt></a>.</li>
<li><strong>cfWf</strong> (<em>CFWriter</em>) &#8211; The <a class="reference internal" href="#libPoMo.cf.CFWriter" title="libPoMo.cf.CFWriter"><tt class="xref py py-class docutils literal"><span class="pre">CFWriter</span></tt></a> object that contains
the VCF files.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">libPomo.cf</a><ul>
<li><a class="reference internal" href="#the-counts-format">The Counts Format</a></li>
<li><a class="reference internal" href="#objects">Objects</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="vcf.html"
                        title="previous chapter">libPomo.vcf</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/cf.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="vcf.html" title="libPomo.vcf"
             >previous</a> |</li>
        <li><a href="index.html">PoMo 1.0.2 documentation</a> &raquo;</li>
          <li><a href="libPoMo.html" >libPoMo</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Dominik Schrempf.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>